### PR TITLE
Prevent cudacodec libs locations being reset when cmake is re-run

### PR DESCRIPTION
**Pull request by cudawarped**
If cmake is run a second time then the location of the Nvidia Video Codec SDK libs are reset due to the implementation of FindCUDA.cmake.  This was observerd in https://github.com/opencv/opencv_contrib/issues/3195#issuecomment-1065351104.

This PR also forces `WITH_NVCUVID` and `WITH_NVCUVENC` to default to `ON` so that if the libs are present decoding and/or encoding support is added to cudacodec without the user having to manually specify it.  This mirrors the way in which the DNN module is built with cuDNN if the cuDNN libs are present without the user having to specify `WITH_CUDNN=ON`.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

```
force_builders=Custom
build_image:Custom=ubuntu-cuda:16.04
buildworker:Custom=linux-1
```

### CODE REVIEW

**Comment by asmorkalov** [comment]
@cudawarped Thanks for the contribution. I suspect that you just need to remove WIN32 condition here: https://github.com/opencv/opencv/blob/4.x/cmake/FindCUDA.cmake#L792 without extra changes in cmake. `find_cuda_helper_libs` calls `find_library` at the end. `find_library` stores results in CMake cache and you should get the expected behavior.

**Comment by cudawarped** [comment]
@asmorkalov Unless an old version of cmake is used or the behaviour is overidden OpenCV use CMake's internal version of FindCUDA which we can't alter.

Inside CMake's version the problem appears to be a result of `CUDA_TOOLKIT_TARGET_DIR` not being set in the cache resulting in `cuda_unset_include_and_libraries()` beng called on every subsequent invocation of cmake [here](https://gitlab.kitware.com/cmake/cmake/-/blob/master/Modules/FindCUDA.cmake#L803)

> if(DEFINED CUDA_TOOLKIT_TARGET_DIR
>     AND NOT "${CUDA_TOOLKIT_TARGET_DIR}" STREQUAL "${CUDA_TOOLKIT_TARGET_DIR_INTERNAL}")
>   cuda_unset_include_and_libraries()
> endif()

My initial fix was to set `CUDA_TOOLKIT_TARGET_DIR` before calling FindCUDA but because changes to FindCUDA are out of our control I thought it would be better to store the current locations of the encoding/decoding libs instead.

Should we instead make changes to the OpenCV version of FindCUDA and force it to be used [here](https://github.com/opencv/opencv/blob/0bd756fde04cc2285934b954d3bab35f24c911e3/cmake/OpenCVDetectCUDA.cmake#L22)

**Comment by asmorkalov** [comment]
@alalek What is your opinion? The proposed solution with `_internal_`  variables works, but looks ugly a bit.

File: cmake/OpenCVDetectCUDA.cmake [hunk 1]
```diff
@@ -75,21 +84,39 @@ if(CUDA_FOUND)
     if(WITH_NVCUVID)
       ocv_cuda_SEARCH_NVCUVID_HEADER("nvcuvid.h" HAVE_NVCUVID_HEADER)
       ocv_cuda_SEARCH_NVCUVID_HEADER("dynlink_nvcuvid.h" HAVE_DYNLINK_NVCUVID_HEADER)
+      if(DEFINED CUDA_nvcuvid_LIBRARY_INTERNAL)
+        set(CUDA_nvcuvid_LIBRARY ${CUDA_nvcuvid_LIBRARY_INTERNAL} CACHE FILEPATH "\"nvcuvid\" library" FORCE)
```
Comment (alalek) [inline cmake/OpenCVDetectCUDA.cmake:88]: > calls to FindCUDA can reset

Should we check for the empty value then?

Comment (cudawarped) [reply cmake/OpenCVDetectCUDA.cmake:88]: @alalek  I agree but I've just noticed this issue was fixed in CMake 6 months ago https://gitlab.kitware.com/cmake/cmake/-/commit/a7758394afc2717bb31a47826e62c1ba86a63c41.  I suggest for compatibility with older versions of CMake we could apply the same fix they used which I mentioned above and simply ensure that `CUDA_TOOLKIT_TARGET_DIR` is set before calling FindCUDA what do you think?

**Review by asmorkalov — APPROVED** [review]

**Pull request merged**